src/syscheckd/whodata/audit_healthcheck.c

```diff
@@ -74,25 +77,35 @@ int audit_health_check(int audit_socket) {
     if (audit_delete_rule(AUDIT_HEALTHCHECK_DIR, WHODATA_PERMS, AUDIT_HEALTHCHECK_KEY) <= 0) {
         mdebug1(FIM_HEALTHCHECK_CHECK_RULE); // LCOV_EXCL_LINE
     }
-    hc_thread_active = 0;
+    atomic_int_set(&hc_thread_active, 0);
+
+    // Lock this thread (with 5 seconds timeout) until the healthcheck thread has ended.
+    w_mutex_lock(&audit_hc_mutex);
+    gettime(&wait_time);
+    wait_time.tv_sec += 5;
+    pthread_cond_timedwait(&audit_hc_started, &audit_hc_mutex, &wait_time);
+    w_mutex_unlock(&audit_hc_mutex);
 
     return retval;
 }
 
 // LCOV_EXCL_START
 void *audit_healthcheck_thread(int *audit_sock) {
-
     w_mutex_lock(&audit_hc_mutex);
-    hc_thread_active = 1;
+    atomic_int_set(&hc_thread_active, 1);
     w_cond_signal(&audit_hc_started);
     w_mutex_unlock(&audit_hc_mutex);
 
     mdebug2(FIM_HEALTHCHECK_THREAD_ACTIVE);
 
-    audit_read_events(audit_sock, HEALTHCHECK_MODE);
+    audit_read_events(audit_sock, &hc_thread_active);
 
     mdebug2(FIM_HEALTHCHECK_THREAD_FINISHED);
 
+    w_mutex_lock(&audit_hc_mutex);
+    w_cond_broadcast(&audit_hc_started);
```

Comment:
I think a `w_cond_signal` makes more sense here, I believe there will be only one thread at most waiting for the condition variable.